src/EnergyPlus/WaterThermalTanks.cc

```diff
@@ -8410,6 +8429,487 @@ namespace WaterThermalTanks {
+        //       RE-ENGINEERED  Noel Merket, November 2018
+
+        // PURPOSE OF THIS SUBROUTINE:
+        // Simulates a stratified, multi-node water heater tank with up to two heating elements.
+
+        // METHODOLOGY EMPLOYED:
+        // This model uses a numerical calculation based on an analytical solution of the ODE dT/dt = a*T + b.
+        // A heat balance is calculated for each node.
+        // Temperatures and energies change dynamically over the system time step.
+        // Final node temperatures are reported as final instantaneous values as well as averages over the
+        // time step.  Heat transfer rates are averages over the time step.
+
+        static std::string const RoutineName("CalcWaterThermalTankStratified");
+        const Real64 TemperatureConvergenceCriteria = 0.0001;
+        const Real64 maxDt = 60.0;
+
+        // Using/Aliasing
+        using DataGlobals::HourOfDay;
+        using DataGlobals::TimeStep;
+        using DataGlobals::TimeStepZone;
+        using DataHVACGlobals::SysTimeElapsed;
+        using DataHVACGlobals::TimeStepSys;
+        using FluidProperties::GetDensityGlycol;
+        using FluidProperties::GetSpecificHeatGlycol;
+
+        // Tank object reference
+        WaterThermalTankData &Tank = WaterThermalTank(WaterThermalTankNum);
+        const Real64 &nTankNodes = Tank.Nodes;
+
+        // Fraction of the current hour that has elapsed (h)
+        const Real64 TimeElapsed = HourOfDay + TimeStep * TimeStepZone + SysTimeElapsed;
+
+        // Seconds in one timestep (s)
+        const Real64 SecInTimeStep = TimeStepSys * SecInHour;
+
+        // Advance tank simulation to the next system timestep, if applicable
+        if (Tank.TimeElapsed != TimeElapsed) {
+            // The simulation has advanced to the next system timestep.  Save conditions from the end of the previous system
+            // timestep for use as the initial conditions of each iteration that does not advance the system timestep.
+            for (auto &e : Tank.Node)
+                e.SavedTemp = e.Temp;
+
+            Tank.SavedHeaterOn1 = Tank.HeaterOn1;
+            Tank.SavedHeaterOn2 = Tank.HeaterOn2;
+
+            // Save outlet temperatures for demand-side flow control
+            Tank.SavedUseOutletTemp = Tank.UseOutletTemp;
+            Tank.SavedSourceOutletTemp = Tank.SourceOutletTemp;
+
+            Tank.TimeElapsed = TimeElapsed;
+        }
+
+        // Reset node temperatures to what they were at the beginning of the system timestep.
+        for (auto &e : Tank.Node)
+            e.Temp = e.SavedTemp;
+
+        // Condenser configuration of heat pump water heater
+        const int HPWHCondenserConfig = Tank.HeatPumpNum > 0 ? HPWaterHeater(Tank.HeatPumpNum).TypeNum : 0;
+
+        // Heat rate from the heat pump (W)
+        const Real64 Qheatpump = [&Tank]{
+            if (!(Tank.HeatPumpNum > 0)) {
+                return 0.0;
+            }
```

Comment:
Why isn't this just    if(Tank.HeatPumpNum == 0) return 0.0;   ??

Also, doesn't this above (line 8464) have some sort of protection where there must be a WaterThermalTank instance? Meaning you wouldn't even need this test at 8499 (i.e., how did we get here if there is no tank)?

    WaterThermalTankData &Tank = WaterThermalTank(WaterThermalTankNum);

Reply:
Good point. I'll change expression in the `if` statement to be more clear.

As for your question about the test, yes, we are guaranteed to have a `Tank` here, but it could be a standalone stratified tank object or it can be connected to a HPWH. This test checks for that and if there's no heat pump there's not heat from the heat pump, which is why we need this check.